### ISSUE TITLE
General Dataset wrapper

### ISSUE THREAD

**Issue by nucli-vicky** [deepinv/deepinv#949]
### Describe the feature

A general `DatasetWrapper` which is able to take a `Dataset` object (with some specified constraints e.g. returns 2D data), and wraps it to add a set of predefined transforms to it. 

Sketch: 
```python
class DatasetWrapper:
   def __init__(self, dataset, ...):
       self.ds = dataset
       # some other things dependent on args
   def __getitem__(self, idx):
       im = self.ds.__getitem__(idx)
       # apply whatever to the im
       return im
```

### Motivation

- Most datasets do not have the same shape for each and every image. `deepinv` (and the batch nature of deep learning implementation) do require that at the end of processing pipelines we do arrive at a set of images with the same shape. 

- Additionally, data augmentation is also an important part of deep learning, which right now `deepinv` does not provide specific support for.

As it is, we require the user to define all their own `transforms` or provide their own `dataset` object with all these factors included. However in a lot of places we are running into situations where we want to add for example some cropping / padding (#913 #948 , ...).

I envision we can provide a nice, well documented set of preprocessing transforms which are suitable to inverse problems (for instance `resize` is often used in natural imaging preprocessing, but it's suitability for inverse problems is up to debate 🙃).
This then allows users to make well informed choices, and also removes a bit of the burden for them to handle of this themselves. 

The other benefit is that this means that we can say for quite some scenario's: "hey, if you use this set of transforms, you're fine and the code will run". For instance in #948 this would be nice and then at least we can link to a doc explaining pro's and con's of things instead of just offering a default with no explanation.

### Additional context

1. We can ofcourse do these things specifically for each scenario (`generate_dataset`, validation, training, ...), but this would lead to duplicate code, difficulties in maintainability, and potential for differences in definition and surprising behavior differences.
2. Instead of a wrapper, we could provide these transforms as is in a way they can be used with the `deepinv` datasets. However, a wrapper can support arbitrary user defined datasets as well; and is agnostic to all things which happened before.
3. One thing I considered for #948 was to make a copy of the dataset object and modify the `transform_x` in place, but this is only fine when `x_transform` is identity (otherwise we remove transforms defined by the user which is not nice behavior), is difficult to detect (we'd have to create an identity function which we default to or use bytecode decomposition, and even then we still (i) do not support user datasets which might not have `x_transform` and (ii) still break for all cases where the user did pass some transform..

**Comment by jscanvic** [comment]
Thanks for raising this point Vicky!

I agree that we need clearer pathways for the use of transforms in datasets. This is actually something that we discussed with Andrew recently and we realized that the transforms module can be used for quite different things that have quite different requisites
- Pre-processing
- Data-augmentation
- Batching
- Use in algorithms (mostly Equivariant Imaging/Splitting/Denoisers/Reconstructors)

I'm not entirely sure how best to account for all these use cases but I assume we're gonna have to split things up at some point.

That being said, I'm not entirely sure I understand the use case for the dataset wrapper you suggest. As far as I know, each dataset holds transform attributes, at least since @Andrewwango introduced the base dataset class. I think it is what you're getting at in the point 3 of Additional context but I'm still not sure what you have in mind. Could you elaborate a bit more on that?

**Comment by Andrewwango** [comment]
Thanks for raising this Vicky! I'm thinking of torchio's nice transforms class, which is very drag and drop with their datasets. On the other hand, I'm reluctant with heavier wrappers - the torchio Subject means it's a bit of a faff to use it in existing torch workflows. 

Either way it would be good to standardise both interfaces for  transforms and also what the "transform" parameter in datasets does. But hopefully it doesn't require a wrapper, just some refactoring! Could be wrong though

**Comment by nucli-vicky** [comment]
The main reason for a wrapper like sketched above, is that we cannot assume the origin of a `Dataset` object and that it might be a custom one which does not have `transform_x` et al., and so the advantage of such a wrapper (not 100% sure if it's the correct term?) would be that we are agnostic to the dataset's origin and not put any requirements or assumptions on which attributes it has (aside from returning e.g. Tensor / array / PIL in format C, H, W), which allows for more flexibility

**Comment by jscanvic** [comment]
Makes sense, thanks for clarifying! My only fear is a "global" wrapper like that will be very abstract on its own. Maybe we could have specific wrappers where it's needed?

This is what I tend to do in my own code. Lately I wanted to access dataset indices from my batches and added a wrapper that returns idx, x, y instead of just x, y.

At least, maybe it can be something internal/private?

**Comment by fepegar** [comment]
> Thanks for raising this Vicky! I'm thinking of torchio's nice transforms class, which is very drag and drop with their datasets. On the other hand, I'm reluctant with heavier wrappers - the torchio Subject means it's a bit of a faff to use it in existing torch workflows.
> 
> Either way it would be good to standardise both interfaces for transforms and also what the "transform" parameter in datasets does. But hopefully it doesn't require a wrapper, just some refactoring! Could be wrong though

Hey, @Andrewwango. Just wanted to mention that TorchIO transforms can take more generic inputs such as a PyTorch tensor or NumPy array, or a `dict` thereof :)

**Comment by Andrewwango** [comment]
Thank you @fepegar for the comment, great to hear it from the source!

In which case, we could have a nicer modular approach with the torchio transforms and unwrapped datasets to keep a simpler interface.

Another idea: it would also be great to eventually have the torchio-style buffered/queued loader - in imaging, we often do 3D patched training (related #806).

**Comment by fepegar** [comment]
FYI I'm playing around with the idea of re-writing TorchIO from scratch, with a similar API but many improvements. A new, much more efficient queue will be one of them.
